Add frequency number handlebars

Kuuuube opened the pull request:
A simple frequency handlebar is something that I think yomitan has been in desperate need for. The current frequencies handlebar is useless for the vast majority of users. Many users are using default handlebars for everything else but have to muck around with custom handlebars to do frequency sorting.

This adds a handlebar that outputs frequency as an integer number. It combines frequency by taking the harmonic mean of all frequencies. This seems to be generally agreed on as the best way to merge frequencies together. When the same dictionary has multiple frequencies, only the first frequency is used.

Based on https://github.com/MarvNC/JP-Resources?tab=readme-ov-file#freq-handlebar. Modified to be more "general use".

I'm not sure if this is an acceptable way to implement this and there are some potential pitfalls but I think this handlebar (or another implementation of it) is very needed and wanted.

Opening this PR as a draft since I'm not familiar enough with the yomitan codebase to tell for sure if I need to touch anything else to properly implement a new handlebar.

toasted-nutbread commented on ext/data/templates/default-anki-field-templates.handlebars:
```diff
@@ -302,6 +302,54 @@
     {{~/if~}}
 {{/inline}}
 
+{{#*inline "frequencies-num"}}
```
Avoid abbreviating the name, `frequencies-number` or `frequency-number` is fine.

toasted-nutbread replied:
(If this is outputting a single number, prefer the singular.)

Kuuuube replied:
Good suggestions, fixed this.

toasted-nutbread commented on ext/data/templates/default-anki-field-templates.handlebars:
```diff
@@ -302,6 +302,54 @@
+            {{/if~}}
+
+            {{~#if (get "read-freq") ~}}
+                {{~#set "numericFrequencyMatch"}}{{~#regexMatch "\d+" ""}}{{~this.frequency~}}{{/regexMatch~}}{{/set~}}
```
Consistent naming scheme for variables. Looks like the precedent is `camelCase` for `get/set` in the current templates, so should probably go with that.

toasted-nutbread commented:
`test/utilities/anki.js` => `getFieldMarkers` => add entry to `terms` and rebuild the test data. `npm run test-code-write`

Kuuuube commented:
> `test/utilities/anki.js` => `getFieldMarkers` => add entry to `terms` and rebuild the test data. `npm run test-code-write`

I've added this. However, there seems to be some issues with the pitch accent rendering in this test unrelated to my PR.

```
        "pitch-accents": "{pitch-accents-render-error}",
        "pitch-accent-graphs": "{pitch-accent-graphs-render-error}",
        "pitch-accent-positions": "{pitch-accent-positions-render-error}",
```

I've excluded these issues from my commit.

I bisected this issue to commit 7a0982b.

Removing the catch that writes `{${marker}-render-error}` in `/ext/js/data/anki-note-builder.js` throws the following error:

```
 FAIL  test/dictionary-data.write.js > Write dictionary data expected data
ReferenceError: document is not defined
 ❯ Module.createPronunciationText ext/js/display/sandbox/pronunciation-generator.js:31:23
     29|     const nasalPositionsSet = nasalPositions.length > 0 ? new Set(nasalPositions) : null;
     30|     const devoicePositionsSet = devoicePositions.length > 0 ? new Set(devoicePositions) : null;
     31|     const container = document.createElement('span');
       |                       ^
     32|     container.className = 'pronunciation-text';
     33|     for (let i = 0, ii = morae.length; i < ii; ++i) {
```

toasted-nutbread commented:
Ooops, broke something again. Check #607, if you temporarily apply that change and then run the write file, it will work as expected. Just don't commit changes to that file in this PR.

toasted-nutbread commented:
Although I guess you already did basically that. So that aside, feel free to mark the PR as ready for review, IMO it's fine to add afterwards.

Kuuuube commented:
Testing #607 it generates exactly the same for `anki-note-builder-test-results.json` as what I put in that previous commit. Should I revert any changes there or is it okay to leave as is?

StefanVukovic99 commented on ext/data/templates/default-anki-field-templates.handlebars:
```diff
@@ -302,6 +302,54 @@
+        {{~set "previousFrequencyDictionary" "" ~}}
+        {{~set "total" 1 ~}}
+
+        {{~#each definition.frequencies~}}
```
seems like a lot of logic for handlebars, maybe do this in `anki-note-data-creator.js`?

StefanVukovic99 commented:
Does this handle both rank-based and occurence-based frequency dicts?

Kuuuube commented:
> Does this handle both rank-based and occurence-based frequency dicts?

It does. But using the harmonic mean may not produce the most desirable results in that case. I can add a normal average handlebar as well. So we have something like `frequency-harmonic` and `frequency-average`.

Kuuuube commented:
I've moved most of the logic to js now. I've added a handlebar for frequency average for users of occurance based frequency dicts. And since occurance based and rank based frequency have opposite bounds for what is considered "nonexistent frequency" I've made a separate handlebar for each there.

StefanVukovic99 commented:
Hate to nitpick, but it's occurEnce.

Kuuuube commented:
> Hate to nitpick, but it's occurEnce.

Oops. All good, important to get things like that right.

toasted-nutbread commented on ext/js/data/sandbox/anki-note-data-creator.js:
```diff
@@ -340,6 +344,68 @@ function getKanjiFrequencies(dictionaryEntry) {
+            continue;
+        }
+        previousDictionary = dictionary;
+        length += 1;
```
`++`

toasted-nutbread commented on ext/js/data/sandbox/anki-note-data-creator.js:
```diff
@@ -340,6 +344,68 @@ function getKanjiFrequencies(dictionaryEntry) {
+
+        const frequencyWorking = displayValue !== null ? displayValue.match(/\d+/) : frequency;
+        if (frequencyWorking !== null) {
+            total += Number(frequencyWorking);
```
Repeated code between this and the last one, make a helper method to get the computed integer value of a frequency.

```js
function getFrequencyNumber(value, displayValue) {
  if (displayValue !== null) {
    const match = displayValue.match(/\d+/);
    if (match !== null) { return Number.parseInt(match[0], 10); }
  }
  return value;
}
```

toasted-nutbread commented on ext/js/data/sandbox/anki-note-data-creator.js:
```diff
@@ -340,6 +344,68 @@ function getKanjiFrequencies(dictionaryEntry) {
+            total += frequency;
+        }
+    }
+    total = Math.floor(total / length);
```
return the math.floor value instead of assigning and returning

toasted-nutbread commented on ext/js/data/sandbox/anki-note-data-creator.js:
```diff
@@ -340,6 +344,68 @@ function getKanjiFrequencies(dictionaryEntry) {
+        return -1;
+    }
+
+    for (const {dictionary, frequency, displayValue} of dictionaryEntry.frequencies) {
```
See comments for below function.

Kuuuube commented:
Got rid of most of the duplicated code. Having two different functions that do the exact same thing between the term and kanji cards seemed weird too so I merged them together.

Looking at getFrequencyAverage and getFrequencyHarmonic right next to eachother I don't love that they're almost the exact same thing but not sure if I should go and crunch those down into another helper function as well.

toasted-nutbread commented on ext/js/data/sandbox/anki-note-data-creator.js:
```diff
@@ -170,6 +170,65 @@ function getPublicContext(context) {
+        }
+        previousDictionary = dictionary;
+
+        const frequencyWorking = displayValue !== null ? displayValue.match(/\d+/) : frequency;
```
IMO this particular logic should still be changed to something more similar to what I mentioned before, since what's happening is actually a bit subtle.

The `Number(frequencyWorking)` conversion does one of two things:
1. Converts the value that is already a number (`frequency`) to a number, so a no-op.
2. Converts the regex match result to a number, but the way that it does this is a bit odd, and it only works due to the regex being formatted the way it is

   `frequencyWorking` will be a `RegExpMatchArray`, which is a special array with custom properties attached to it. The custom properties aren't relevant, but the fact that it's an array is. Passing this into the `Number` function will first convert the argument to a string. For arrays, this stringifies each element and concatenates them with `','`. As it turns out, the `RegExpMatchArray`'s length is `1` because the regex has no capture groups. but consider the following:

   `Number('1234'.match(/(\d+)/))` yields `NaN` because it is effectively doing `Number('1234,1234')`
   
   So what I think would be more clear type safety would be to do something similar to what I mentioned in a previous comment with the `getFrequencyNumber` function. The reason why it being in a separate function helps is because it allows you to just do `frequencies.push(getFrequencyNumber(the, arguments))`, which only takes one `.push` call and is a bit easier to do with a `return` in a function since it's nested two if's deep.


Additionally, I recommend either `parseFloat(v)` or `parseInt(v, 10)` due to the scope of usage being more explicitly limited to string conversions, and also because those are the more commonly used functions throughout the codebase.

Please enjoy this table of JavaScript madness:
| expression | result |
| --- | --- |
| `Number.parseInt('1234,1234', 10)` | `1234` |
| `Number.parseFloat('1234,1234')` | `1234` |
| `Number('1234,1234')` | `NaN` |

---

Now, one could argue that none of this matters because it works as it is and performance is irrelevant for this function, but I am pedantic.

Kuuuube replied:
Well that's insane... 

I've changed this to use parseInt now. I quite like just being able to toss the dict info to `getFrequencyNumbers` and getting back the nice data instead of having to do any checks over the data in `getFrequencyHarmonic` or `getFrequencyAverage`. 

Using `continue` in the loop kindof does what `return` would do if I had split it out exactly like you said so that hopefully cleans up that well enough.

toasted-nutbread commented on ext/js/data/sandbox/anki-note-data-creator.js:
```diff
@@ -170,6 +170,65 @@ function getPublicContext(context) {
+function getFrequencyHarmonic(dictionaryEntry) {
+    const frequencies = getFrequencyNumbers(dictionaryEntry);
+
+    if (frequencies.length < 1) {
```
`length === 0`

toasted-nutbread commented:
IMO it's not worth trying to crush it even further. If you want smaller code, just collapse what you have currently.
```js
    const frequencies = getFrequencyNumbers(dictionaryEntry);
    if (frequencies.length < 1) { return -1; }
    let total = 0;
    for (const frequency of frequencies) { total += 1 / frequency; }
    return Math.floor(frequencies.length / total);
```
